pkg/udf/forward/forward.go

```diff
@@ -399,12 +371,77 @@ func (isdf *InterStepDataForward) forwardAChunk(ctx context.Context) {
+func (isdf *InterStepDataForward) processConcurrentMap(ctx context.Context, dataMessages []*isb.ReadMessage) []isb.ReadWriteMessagePair {
+	udfResults := make([]isb.ReadWriteMessagePair, len(dataMessages))
+	// udf concurrent processing request channel
+	udfCh := make(chan *isb.ReadWriteMessagePair)
+	// udfResults stores the results after map UDF processing for all read messages. It indexes
+	// a read message to the corresponding write message
+	// applyUDF, if there is an Internal error it is a blocking call and will return only if shutdown has been initiated.
+
+	// create a pool of map UDF Processors
+	var wg sync.WaitGroup
+	for i := 0; i < isdf.opts.udfConcurrency; i++ {
+		wg.Add(1)
+		go func() {
+			defer wg.Done()
+			isdf.concurrentApplyUDF(ctx, udfCh)
+		}()
+	}
+	concurrentUDFProcessingStart := time.Now()
+
+	// send to map UDF only the data messages
+	for idx, m := range dataMessages {
+		// emit message size metric
+		metrics.ReadBytesCount.With(map[string]string{metrics.LabelVertex: isdf.vertexName, metrics.LabelPipeline: isdf.pipelineName, metrics.LabelVertexType: string(dfv1.VertexTypeMapUDF), metrics.LabelVertexReplicaIndex: strconv.Itoa(int(isdf.vertexReplica)), metrics.LabelPartitionName: isdf.fromBufferPartition.GetName()}).Add(float64(len(m.Payload)))
+		// send map UDF processing work to the channel
+		udfResults[idx].ReadMessage = m
+		udfCh <- &udfResults[idx]
+	}
+	// let the go routines know that there is no more work
+	close(udfCh)
+	// wait till the processing is done. this will not be an infinite wait because the map UDF processing will exit if
+	// context.Done() is closed.
+	wg.Wait()
+	isdf.opts.logger.Debugw("concurrent applyUDF completed", zap.Int("concurrency", isdf.opts.udfConcurrency), zap.Duration("took", time.Since(concurrentUDFProcessingStart)))
+	metrics.ConcurrentUDFProcessingTime.With(map[string]string{metrics.LabelVertex: isdf.vertexName, metrics.LabelPipeline: isdf.pipelineName, metrics.LabelVertexType: string(dfv1.VertexTypeMapUDF), metrics.LabelVertexReplicaIndex: strconv.Itoa(int(isdf.vertexReplica))}).Observe(float64(time.Since(concurrentUDFProcessingStart).Microseconds()))
+	return udfResults
+}
+
+// processBatchMessages is used for processing the Batch Map mode UDF
+// batch map processing we send a list of N input requests together to the UDF and get the consolidated
+// response for all of them.
+func (isdf *InterStepDataForward) processBatchMessages(ctx context.Context, dataMessages []*isb.ReadMessage) []isb.ReadWriteMessagePair {
+	concurrentUDFProcessingStart := time.Now()
+	udfResults, err := isdf.mapAppliers.BatchMapUDF.ApplyBatchMap(ctx, dataMessages)
+	if err != nil {
+		// In case of an error received while processing the UDF call, we would do not handle partial failures,
+		// and hence would want to replay the whole batch currently.
+		// For achieving this, no-ack all the messages in the batch to force a re-read from the ISB
```

Comment:
this cannot be held true all the time, ack can fail midway too. please add in the documentation that Batch Map can behave both way based on where the failure is.